Dynamic bundle assets

garak opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no (new method in interface as annotation)
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #29213 
| License       | MIT
| Doc PR        | none (yet)

Everything is explained in linked issue

derrabus commented:
This PR should target the 4.4 branch. If you only add a `@method` annotation to the `BundleInterface` instead of the real method, it's not a BC break anymore.

derrabus commented on src/Symfony/Component/HttpKernel/Bundle/BundleInterface.php:
```diff
@@ -68,4 +68,12 @@ public function getNamespace();
+     *
+     * @return string The relative path for public assets
+     */
+    public function getPublicPath();
```
`: string`

garak replied:
Done

garak commented:
> This PR should target the 4.4 branch. If you only add a `@method` annotation to the `BundleInterface` instead of the real method, it's not a BC break anymore.

So, should I close this PR and open another one on 4.4?

fabpot commented:
@garak No, you change the target of this PR without creating a new one.

garak commented:
@fabpot I just tried to change target to 4.4. and a bunch of stuff popped up.
Maybe a new PR is a better option... unless you can suggest a way to get rid of such mess
Thanks

fabpot commented:
@garak rebase your PR locally on 4.4 and push again, that should resolve the mess.

garak committed:
Dynamic bundle assets

garak commented:
I rebased upon 4.4.
I followed suggestion from @derrabus and replaced actual method with an annotation inside interface.
I also updated initial PR comment.

Hope everything is fine

jvasseur commented on src/Symfony/Bundle/FrameworkBundle/Command/AssetsInstallCommand.php:
```diff
@@ -137,7 +137,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
         /** @var BundleInterface $bundle */
         foreach ($kernel->getBundles() as $bundle) {
-            if (!is_dir($originDir = $bundle->getPath().'/Resources/public')) {
+            if (!is_dir($originDir = $bundle->getPath().$bundle->getPublicPath())) {
```
This can be a BC break in case someone implements the `BundleInterface` from scratch instead of extending the `Bundle` class.

garak replied:
@jvasseur that was what I thinked at first, when I pointed this PR to Symfony 5.

derrabus replied:
But you can check with `method_exists()` if the new method is present. If it isn’t, fall back to `/Resources/public` and trigger a deprecation warning.

garak replied:
Are you sure deprecation is a good idea? Inside a loop, one would get N warnings (where N = number of bundles) for a single deprecation

derrabus replied:
Most bundles use the `Bundle` class anyway, so it shouldn't be that many. We could technically collect all bundles that don't implement the method and issue one warning at the end, but I don't think that this is really necessary.

Anyway, a deprecation warning needs to happen one way or another because the user needs to be made aware of that fact that certain bundles need to be updated.

garak replied:
Also done here

garak commented:
I still see "Status: Needs Work" label. Should I do something else?

derrabus commented on src/Symfony/Component/HttpKernel/Bundle/Bundle.php:
```diff
@@ -135,6 +135,11 @@ public function registerCommands(Application $application)
     {
     }
 
+    public function getPublicPath()
```
```suggestion
    public function getPublicPath(): string
```

garak commented:
Friendly reminder: did needed work, do I need to do some more?

fabpot commented on src/Symfony/Component/HttpKernel/Bundle/Bundle.php:
```diff
@@ -135,6 +135,11 @@ public function registerCommands(Application $application)
 
+    public function getPublicPath(): string
+    {
+        return '/Resources/public';
```
Same here

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Command/AssetsInstallCommand.php:
```diff
@@ -137,7 +137,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
-            if (!is_dir($originDir = $bundle->getPath().'/Resources/public')) {
+            if (!method_exists($bundle, 'getPublicPath')) {
+                @trigger_error('Not definind getPublicPath() method is deprecated since Symfony 4.4 and will not be supported in 5.0.', E_USER_DEPRECATED);
+                $publicPath = '/Resources/public';
```
If it's a relative path, I would not use `/` at the beginning of the value.

garak replied:
This was the old value (I mean, the value that is currently returned in Symfony code), I think we should keep as it

fabpot replied:
Where is it returned?

garak replied:
The value currently used by Symfony for `$originDir` is `$bundle->getPath().'/Resources/public'`, as you can see above, in line 140

fabpot replied:
Which is different, right? Here, we have a relative path, there, an absolute one.

garak replied:
Sorry, I don't get you.
The old code is `$originDir = $bundle->getPath().'/Resources/public'`
The new code is `$publicPath = '/Resources/public'; $originDir = $bundle->getPath().$publicPath` (except for checking the new method, of course)
I can't see any change about relative/absolute.

fabpot replied:
I think that the code should be `$publicPath = 'Resources/public';` and then `if (!is_dir($originDir = $bundle->getPath().'/'.$publicPath)) {`

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Command/AssetsInstallCommand.php:
```diff
@@ -137,7 +137,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
         foreach ($kernel->getBundles() as $bundle) {
-            if (!is_dir($originDir = $bundle->getPath().'/Resources/public')) {
+            if (!method_exists($bundle, 'getPublicPath')) {
+                @trigger_error('Not definind getPublicPath() method is deprecated since Symfony 4.4 and will not be supported in 5.0.', E_USER_DEPRECATED);
```
typo: `defining`

`Not defining the "getPublicPath()" method...`

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Command/AssetsInstallCommand.php:
```diff
@@ -137,7 +137,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
+            } else {
+                $publicPath = $bundle->getPublicPath();
+            }
+            if (!is_dir($originDir = $bundle->getPath().$publicPath)) {
```
As the path is relative, you should add `DIRECTORY_PATH` between getPath() and $publicPath.

fabpot reviewed (changes requested):
Having some tests for this new feature would be nice as well.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Command/AssetsInstallCommand.php:
```diff
@@ -138,12 +138,12 @@ protected function execute(InputInterface $input, OutputInterface $output)
             if (!method_exists($bundle, 'getPublicPath')) {
-                @trigger_error('Not definind getPublicPath() method is deprecated since Symfony 4.4 and will not be supported in 5.0.', E_USER_DEPRECATED);
-                $publicPath = '/Resources/public';
+                @trigger_error('Not defining getPublicPath() method is deprecated since Symfony 4.4 and will not be supported in 5.0.', E_USER_DEPRECATED);
```
`"getPublicPath()"`

fabpot commented on src/Symfony/Component/HttpKernel/Tests/KernelTest.php:
```diff
@@ -649,6 +649,13 @@ protected function getBundle($dir = null, $parent = null, $className = null, $bu
+            ->willReturn('Resources/public')
+        ;
+
+
```
extra white line here could be removed.

garak commented:
I addressed all notes from @fabpot, I guess. Some portions of code are not shown as modified, but they are, actually.
I also added a small expectation in tests, I couldn't find any direct tests of AssetsInstallCommand. Failures in CI are unrelated

fabpot commented:
Thank you @garak.

Tobion commented on src/Symfony/Component/HttpKernel/Bundle/Bundle.php:
```diff
@@ -135,6 +135,11 @@ public function registerCommands(Application $application)
 
+    public function getPublicPath(): string
+    {
+        return 'Resources/public';
```
IMO that should return the absolute dir to be consistent with getLogDir etc, 

https://github.com/symfony/symfony/blob/08c80eb3419606fa6f605037d5f6a392187f90c7/src/Symfony/Component/HttpKernel/Kernel.php#L346

garak replied:
@Tobion here we're not in the project context, but in the bundle one. Anyway, bundle always returned such path, this is not changing anything
